File: core/src/main/scala/com/github/propi/rdfrules/data/Graph.scala
package com.github.propi.rdfrules.data

import java.io._
import com.github.propi.rdfrules.algorithm.{RuleConsumer, RulesMining}
import com.github.propi.rdfrules.algorithm.consumer.InMemoryRuleConsumer
import com.github.propi.rdfrules.data.Quad.QuadTraversableView
import com.github.propi.rdfrules.data.Triple.TripleTraversableView
import com.github.propi.rdfrules.data.ops._
import com.github.propi.rdfrules.index.Index
import com.github.propi.rdfrules.ruleset.Ruleset
import com.github.propi.rdfrules.serialization.TripleSerialization._
import com.github.propi.rdfrules.utils.{Debugger, ForEach}
import com.github.propi.rdfrules.utils.serialization.{Deserializer, SerializationSize, Serializer}

import scala.language.implicitConversions
import scala.reflect.ClassTag

/**
  * Created by Vaclav Zeman on 3. 10. 2017.
  */
class Graph private(val name: TripleItem.Uri, val triples: TripleTraversableView, val userDefinedPrefixes: ForEach[Prefix])
  extends Transformable[Triple, Graph]
    with TriplesOps
    with QuadsOps[Graph]
    with PrefixesOps[Graph]
    with Discretizable[Graph]
    with Cacheable[Triple, Graph]
    with Debugable[Triple, Graph]
    with Sampleable[Triple, Graph] {

  protected val serializer: Serializer[Triple] = implicitly[Serializer[Triple]]
  protected val deserializer: Deserializer[Triple] = implicitly[Deserializer[Triple]]
  protected val serializationSize: SerializationSize[Triple] = implicitly[SerializationSize[Triple]]
  protected val dataLoadingText: String = "Graph loading"

  protected def coll: TripleTraversableView = triples

  protected def cachedTransform(col: TripleTraversableView): Graph = new Graph(name, col, userDefinedPrefixes)

  protected def transform(col: TripleTraversableView): Graph = new Graph(name, col, userDefinedPrefixes)

  protected def transformQuads(col: QuadTraversableView): Graph = transform(col.map(_.triple))

  protected def transformPrefixesAndColl(prefixes: ForEach[Prefix], col: QuadTraversableView): Graph = new Graph(name, col.map(_.triple), prefixes)

  protected def valueClassTag: ClassTag[Triple] = implicitly[ClassTag[Triple]]

  override protected def samplingDistributor: Option[Triple => Any] = Some(_.predicate)

  def foreach(f: Triple => Unit): Unit = triples.foreach(f)

  def `export`(os: => OutputStream)(implicit writer: RdfWriter): Unit = writer.writeToOutputStream(this, os)

  def `export`(file: File)(implicit writer: RdfWriter): Unit = {
    val newWriter = if (writer == RdfWriter.NoWriter) RdfWriter(file) else writer
    `export`(new FileOutputStream(file))(newWriter)
  }

  def `export`(file: String)(implicit writer: RdfWriter): Unit = `export`(new File(file))

  def quads: QuadTraversableView = triples.map(_.toQuad(name))

  def withName(name: TripleItem.Uri): Graph = new Graph(name, triples, userDefinedPrefixes)

  def toDataset: Dataset = Dataset(this)

  def mine(miner: RulesMining, ruleConsumer: RuleConsumer.Invoker[Ruleset] = RuleConsumer(InMemoryRuleConsumer()))(implicit debugger: Debugger = Debugger.EmptyDebugger): Ruleset = toDataset.mine(miner, ruleConsumer)

  def index(implicit debugger: Debugger = Debugger.EmptyDebugger): Index = toDataset.index()

}

object Graph {

  val default: TripleItem.Uri = TripleItem.LongUri("")

  def apply(triples: TripleTraversableView): Graph = new Graph(default, triples, ForEach.empty)

  def apply(name: TripleItem.Uri, triples: TripleTraversableView): Graph = new Graph(name, triples, ForEach.empty)

  def apply(name: TripleItem.Uri, is: => InputStream)(implicit reader: RdfReader): Graph = new Graph(name, reader.fromInputStream(is).map(_.triple), ForEach.empty)

  def apply(name: TripleItem.Uri, file: File)(implicit reader: RdfReader): Graph = {
    val newReader = if (reader == RdfReader.NoReader) RdfReader(file) else reader
    new Graph(name, newReader.fromFile(file).map(_.triple), ForEach.empty)
  }

  def apply(name: TripleItem.Uri, file: String)(implicit reader: RdfReader): Graph = apply(name, new File(file))

  def apply(file: File)(implicit reader: RdfReader): Graph = apply(default, file)

  def apply(file: String)(implicit reader: RdfReader): Graph = apply(new File(file))

  def apply(is: => InputStream)(implicit reader: RdfReader): Graph = apply(default, is)

  def fromCache(name: TripleItem.Uri, is: => InputStream): Graph = new Graph(
    name,
    (f: Triple => Unit) => {
      Deserializer.deserializeFromInputStream[Triple, Unit](is) { reader =>
        Iterator.continually(reader.read()).takeWhile(_.isDefined).foreach(x => f(x.get))
      }
    },
    ForEach.empty
  )

  def fromCache(is: => InputStream): Graph = fromCache(default, is)

  def fromCache(name: TripleItem.Uri, file: File): Graph = fromCache(name, new FileInputStream(file))

  def fromCache(name: TripleItem.Uri, file: String): Graph = fromCache(name, new File(file))

  def fromCache(file: File): Graph = fromCache(default, file)

  def fromCache(file: String): Graph = fromCache(default, new File(file))

}

File: java/src/main/java/com/github/propi/rdfrules/java/data/Graph.java
package com.github.propi.rdfrules.java.data;

import com.github.propi.rdfrules.java.ReadersWriters;
import com.github.propi.rdfrules.java.ScalaConverters;
import com.github.propi.rdfrules.java.TripleItemConverters;
import com.github.propi.rdfrules.java.algorithm.RulesMining;
import com.github.propi.rdfrules.java.index.Index;
import com.github.propi.rdfrules.java.ruleset.Ruleset;
import org.apache.jena.riot.Lang;
import org.apache.jena.riot.RDFFormat;

import java.io.File;
import java.io.InputStream;
import java.io.OutputStream;
import java.util.function.Consumer;
import java.util.function.Supplier;

/**
 * Created by Vaclav Zeman on 3. 5. 2018.
 */
public class Graph implements
        Transformable<com.github.propi.rdfrules.data.Triple, Triple, com.github.propi.rdfrules.data.Graph, Graph>,
        Cacheable<com.github.propi.rdfrules.data.Triple, com.github.propi.rdfrules.data.Graph, Graph>,
        QuadsOps<com.github.propi.rdfrules.data.Graph, Graph>,
        Discretizable<com.github.propi.rdfrules.data.Graph, Graph>,
        TriplesOps {

    final private com.github.propi.rdfrules.data.Graph graph;

    public Graph(com.github.propi.rdfrules.data.Graph graph) {
        this.graph = graph;
    }

    final public static TripleItem.Uri DEFAULT = new TripleItem.LongUri("");

    public static Graph fromTsv(Supplier<InputStream> isb) {
        return new Graph(com.github.propi.rdfrules.data.Graph.apply(isb::get, ReadersWriters.tsvReader()));
    }

    public static Graph fromTsv(TripleItem.Uri name, Supplier<InputStream> isb) {
        return new Graph(com.github.propi.rdfrules.data.Graph.apply(name.asScala(), isb::get, ReadersWriters.tsvReader()));
    }

    public static Graph fromTsv(File file) {
        return new Graph(com.github.propi.rdfrules.data.Graph.apply(file, ReadersWriters.tsvReader()));
    }

    public static Graph fromTsv(TripleItem.Uri name, File file) {
        return new Graph(com.github.propi.rdfrules.data.Graph.apply(name.asScala(), file, ReadersWriters.tsvReader()));
    }

    public static Graph fromTsv(String file) {
        return fromTsv(new File(file));
    }

    public static Graph fromTsv(TripleItem.Uri name, String file) {
        return fromTsv(name, new File(file));
    }

    public static Graph fromRdfLang(Supplier<InputStream> isb, Lang lang) {
        return new Graph(com.github.propi.rdfrules.data.Graph.apply(isb::get, ReadersWriters.jenaLangReader(lang)));
    }

    public static Graph fromRdfLang(TripleItem.Uri name, Supplier<InputStream> isb, Lang lang) {
        return new Graph(com.github.propi.rdfrules.data.Graph.apply(name.asScala(), isb::get, ReadersWriters.jenaLangReader(lang)));
    }

    public static Graph fromRdfLang(File file, Lang lang) {
        return new Graph(com.github.propi.rdfrules.data.Graph.apply(file, ReadersWriters.jenaLangReader(lang)));
    }

    public static Graph fromRdfLang(TripleItem.Uri name, File file, Lang lang) {
        return new Graph(com.github.propi.rdfrules.data.Graph.apply(name.asScala(), file, ReadersWriters.jenaLangReader(lang)));
    }

    public static Graph fromRdfLang(String file, Lang lang) {
        return fromRdfLang(new File(file), lang);
    }

    public static Graph fromRdfLang(TripleItem.Uri name, String file, Lang lang) {
        return fromRdfLang(name, new File(file), lang);
    }

    public static Graph fromFile(File file) {
        return new Graph(com.github.propi.rdfrules.data.Graph.apply(file, ReadersWriters.noRdfReader()));
    }

    public static Graph fromFile(String file) {
        return fromFile(new File(file));
    }

    public static Graph fromTriples(Iterable<Triple> triples) {
        return new Graph(com.github.propi.rdfrules.data.Graph.apply(ScalaConverters.toIterable(triples, Triple::asScala)));
    }

    public static Graph fromTriples(TripleItem.Uri name, Iterable<Triple> triples) {
        return new Graph(com.github.propi.rdfrules.data.Graph.apply(name.asScala(), ScalaConverters.toIterable(triples, Triple::asScala)));
    }

    public static Graph fromCache(Supplier<InputStream> isb) {
        return new Graph(com.github.propi.rdfrules.data.Graph.fromCache(isb::get));
    }

    public static Graph fromCache(TripleItem.Uri name, Supplier<InputStream> isb) {
        return new Graph(com.github.propi.rdfrules.data.Graph.fromCache(name.asScala(), isb::get));
    }

    public static Graph fromCache(File file) {
        return new Graph(com.github.propi.rdfrules.data.Graph.fromCache(file));
    }

    public static Graph fromCache(TripleItem.Uri name, File file) {
        return new Graph(com.github.propi.rdfrules.data.Graph.fromCache(name.asScala(), file));
    }

    public static Graph fromCache(String file) {
        return fromCache(new File(file));
    }

    public static Graph fromCache(TripleItem.Uri name, String file) {
        return fromCache(name, new File(file));
    }

    @Override
    public com.github.propi.rdfrules.data.Graph asScala() {
        return graph;
    }

    @Override
    public Graph asJava(com.github.propi.rdfrules.data.Graph scala) {
        return new Graph(scala);
    }

    @Override
    public Triple asJavaItem(com.github.propi.rdfrules.data.Triple x) {
        return new Triple(x);
    }

    @Override
    public com.github.propi.rdfrules.data.Triple asScalaItem(Triple x) {
        return x.asScala();
    }

    public TripleItem.Uri getName() {
        return TripleItemConverters.toJavaUri(asScala().name());
    }

    public void forEach(Consumer<Triple> consumer) {
        asScala().triples().foreach(v1 -> {
            consumer.accept(asJavaItem(v1));
            return null;
        });
    }

    public void export(Supplier<OutputStream> osb) {
        asScala().export(osb::get, ReadersWriters.tsvWriter());
    }

    public void export(Supplier<OutputStream> osb, RDFFormat rdfFormat) {
        asScala().export(osb::get, ReadersWriters.jenaLangWriter(rdfFormat));
    }

    public void export(File file) {
        asScala().export(file, ReadersWriters.noRdfWriter());
    }

    public void export(File file, RDFFormat rdfFormat) {
        asScala().export(file, ReadersWriters.jenaLangWriter(rdfFormat));
    }

    public void export(String file) {
        asScala().export(file, ReadersWriters.noRdfWriter());
    }

    public void export(String file, RDFFormat rdfFormat) {
        asScala().export(file, ReadersWriters.jenaLangWriter(rdfFormat));
    }

    public Graph withName(TripleItem.Uri name) {
        return asJava(asScala().withName(name.asScala()));
    }

    public Dataset toDataset() {
        return Dataset.fromGraph(this);
    }

    public Index index(Index.Mode mode) {
        return toDataset().index(mode);
    }

    public Index index() {
        return toDataset().index();
    }

    public Ruleset mine(RulesMining rulesMining) {
        return toDataset().mine(rulesMining);
    }

}